go-controller/pkg/crd/userdefinednetwork/v1/types.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:path=userdefinednetworks
+// +kubebuilder:scope=Namespaced
+// +kubebuilder:shortName=udn
+// +kubebuilder:singular=userdefinednetwork
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+
+// UserDefinedNetwork describe network request for a Namespace
+type UserDefinedNetwork struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// +kubebuilder:validation:Required
+	// +required
+	Spec UserDefinedNetworkSpec `json:"spec"`
+	// +optional
+	Status UserDefinedNetworkStatus `json:"status,omitempty"`
+}
+
+// UserDefinedNetworkSpec defines the desired state of UserDefinedNetworkSpec
+type UserDefinedNetworkSpec struct {
+	// The topological configuration for the network
+	Topology NetworkTopology `json:"topology"`
+
+	// The network role in the pod (e.g.: Primary, Secondary).
+	Role NetworkRole `json:"role"`
+
+	// The maximum transmission unit (MTU).
+	// The default value, 1400, is automatically set by the kernel.
+	// +optional
+	Mtu uint `json:"mtu,omitempty"`
+
+	// The subnet to use for the network across the cluster.
+	// Only include the CIDR for the node. E.g. 10.100.200.0/24.
+	//
+	// IPv6 (2001:DBB::/64) and dual-stack (192.168.100.0/24,2001:DBB::/64) subnets are supported.
+	//
+	// When omitted, the logical switch implementing the network only provides layer 2 communication,
+	// and users must configure IP addresses for the pods.
+	// Port security only prevents MAC spoofing.
+	// +optional
+	Subnets []string `json:"subnets,omitempty"`
+
+	// A list of CIDRs.
+	// IP addresses are removed from the assignable IP address pool and are never passed to the pods.
+	// +optional
+	ExcludeSubnets []string `json:"excludeSubnets,omitempty"`
+
+	// Subnet used inside the OVN network topology.
+	// E.g. 100.65.0.0/24. IPv6 (fd99::/64) and dual-stack (100.65.0.0/24,fd99::/64).
+	// When omitted and network `role` is `primary` the following default CIDR is used `100.65.0.0/24`, IPv6 `fd99::/64`.
```

Comment:
From what I understand the default join subnets will eventually be added to `go-controller/pkg/types/const.go` [1] 
Once the PR is merged we can change the net-attach-def template package to take it from there.

@tssurya could you please refer the original PR for [1]?

[1]  https://github.com/ovn-org/ovn-kubernetes/commit/05e13058d038886aebdf1d28e4f841a7fc0b4889#diff-ee7e8cb2c18f89a20eaaf4c10d14e404559cc7c7dbaabe1801566ba57e2b5ed7

Reply:
what happens if joinSubnets are omitted, but network role is not primary?

Reply:
I think we should say in the comment, that this field is only used by networkRole=Primary, and ignored otherwise. When it is unset for primary network, ovn-k will assign joinSubnet (without saying which subnet exactly it will be, as it is not a part of the API)